### PR TITLE
Add retries for schema registry errors

### PR DESCRIPTION
**Pull request by gordon-rennie**
Initial PR for feedback - aims to fix https://github.com/fd4s/fs2-kafka/issues/663

## Goal
Prevent transient schema registry API errors from blowing up Vulcan ser/des by adding retries with sensible default config.

## Implementation
* some code analysis was require to determine when to retry:
  * schema registry client operations can throw `IOException` or `RestClientException`
  * for serialization, the relevant Java [KafkaAvroSerializer](https://github.com/confluentinc/schema-registry/blob/4cbf07a513f3a2fe999088abb5a74d544707c906/avro-serializer/src/main/java/io/confluent/kafka/serializers/AbstractKafkaAvroSerializer.java#L141-L147) error handling maps both `IOException` and `RestClientException` (if non-client error code) to `SerializationException`
  * for deserialization, the [vulcan code](https://github.com/fd4s/fs2-kafka/blob/0cde0b0e8d227fa115271598dc4fb45102fa1c7c/modules/vulcan/src/main/scala/fs2/kafka/vulcan/AvroDeserializer.scala#L32-L33) makes a call using the schema registry client itself and further calls are made by the underlying Java code ([1](https://github.com/fd4s/fs2-kafka/blob/0cde0b0e8d227fa115271598dc4fb45102fa1c7c/modules/vulcan/src/main/scala/fs2/kafka/vulcan/AvroDeserializer.scala#L40), [2](https://github.com/confluentinc/schema-registry/blob/4cbf07a513f3a2fe999088abb5a74d544707c906/avro-serializer/src/main/java/io/confluent/kafka/serializers/AbstractKafkaAvroDeserializer.java#L106-L107), [3](https://github.com/confluentinc/schema-registry/blob/4cbf07a513f3a2fe999088abb5a74d544707c906/avro-serializer/src/main/java/io/confluent/kafka/serializers/AbstractKafkaAvroDeserializer.java#L305-L320)) which are mapped again to `SerializationException`
  * tl;dr: retry on `IOException`, `RestClientException`, `SerializationException` 
* added retry logic for the above exceptions
* added config for number of retries and backoff duration
  * default config is 5 retries with exponential backoff starting at 50 millis.

## Notes
* Adding backoff to retries requires `Temporal[F]`, which requires upgrading from `F: Sync` to `F: Async`. Presumably this is considered a breaking change and so this feature needs to start targetting 3.x instead? I'm new to library code, just let me know :smile: 
* There is no provision for logging in `fs2-kafka-vulcan` (there's only a private, sealed interface in `fs2-kafka` core) so I have not added any
* Retry logic is hand-crafted rather than using a dependency like `cats-retry`
* I was influenced by discussions in issue https://github.com/fd4s/fs2-kafka/issues/623 not to create any new retry interfaces and not to use `Jitter[F]`, which we seem to want to move away from
  * from the discussions in this issue, I think we would prefer a more generic interface like `def schemaRegistryRetryPolicy[A]: (F[A], Throwable) => F[A]` - this would give good flexibility for clients and allow interop with libraries like `cats-retry`. I really like this idea. Unfortunately I can't see a way to make this work: deserialization will return an arbitrary `F[A]` and `A` is basically a wildcard type that needs to accommodate any arbitrary `vulcan.Codec[A]`. Wildcard types aren't supported for anonymous functions, e.g. I can't do: 
```scala
  private[this] final case class AvroSettingsImpl[F[_]](
    ...
    schemaRegistryRetryPolicy: (F[*], SerializationException) => F[*]
```
  feedback on the interface very welcome!

## TODO
* agree/finalize interface changes
* unit test

### CODE REVIEW

**Comment by gordon-rennie** [comment]
(rebasing from 2.x to 3.x had issues and conflicts, I just applied the diff patch as a patch instead before force-pushing but GitHub hasn't liked that. I will re-raise the PR against 3.x)
